Replace rpmbuild dependency with new rpmwriter module

giuseppe opened the pull request:
## Description

Replace rpmbuild dependency with a simpler module to write directly an RPM file.

We lose the capability of building an RPM from a spec file, but this was really an over-engineered solution, so better drop it before it is too late.

Users who want a more complex RPM file, then they must provide it as part of the image itself '/exports/container.rpm'


## Related Issue Numbers

https://github.com/projectatomic/atomic/issues/981

giuseppe commented:
@cgwalters @baude this is the WIP for replacing rpmbuild, I need to add tests

giuseppe commented:
I've adapted the tests to work with the new module.  There are quite a few tests to cover the rpm generation. Dropping the WIP and ready for review

giuseppe commented:
The CI failure is unrelated to the PR

ashcrow commented:
```
+ '[' /var/mnt /sysroot/ostree/deploy/centos-atomic-host/var/mnt ']'
./tests/integration/test_storage.sh: line 39: [: /var/mnt: unary operator expected
+ wipefs -a /dev/vdb
wipefs: error: /dev/vdb: probing initialization failed: Device or resource busy
```

ashcrow commented:
Related: https://github.com/ostreedev/ostree/pull/859

baude commented:
@giuseppe assuming https://github.com/projectatomic/atomic/pull/1008 merges, just rebase and you should be good.

giuseppe commented:
rebased on master so that the storage test is disabled ⬆️

cgwalters commented:
How did you come up with `rpmwriter.py`?  Did you just reverse-engineer librpm and translate to Python?  A few comments about how you did it would help maintenance down the line.

Down the line...I'd like to change things so that the "synthesize an RPM" code lives in `rpm-ostree` in that case.  Basically have atomic pass the metadata as JSON/GVariant/whatever  + path to rootfs.

For now though this seems pretty sane to me.

giuseppe commented:
@cgwalters I've used the documentation here as a starting point: http://ftp.rpm.org/max-rpm/s1-rpm-file-format-rpm-file-format.html

I got stuck a few times so I added more debug messages to `rpm` that helped me, but in general the file format documentation + [rpmtag.h](https://github.com/rpm-software-management/rpm/blob/master/lib/rpmtag.h) was enough to write `rpmwriter.h`.  I've added a fixup patch to add more info about it

TomSweeneyRedHat commented on Atomic/rpmwriter.py:
```diff
@@ -0,0 +1,369 @@
+    RPMTAG_REQUIRENAME		= 1049
+    RPMTAG_REQUIREVERSION	= 1050
+    RPMTAG_CONFLICTFLAGS	= 1053
+    RPMTAG_CONFLICTNAME		= 1054
```
micro nit.  You've an extra tab it looks like between the variable and the equals for a few of these definitions.    I personally like this style and would ask that you do the same for the group of variables above too (lines 40->62)

giuseppe replied:
I've pushed a fixup patch to use the same style for all variables

TomSweeneyRedHat commented on Atomic/rpmwriter.py:
```diff
@@ -0,0 +1,369 @@
+
+    def add_header(self, tag, typ, count, value, pad=1):
+        try:
+            # this fails on python3
```
Should a TODO comment be added to fix in python3?

giuseppe replied:
no, this is the desired behavior.  The try inhibits the exception to be raised as there is no `unicode` in python3

TomSweeneyRedHat replied:
Ah, I read it to mean you were having unexpected issues here.  Perhaps change the comment to something like:

"The next line is expected to always fail under python3, in that case just continue."

or something similar.  Another nit, not a hold the merge comment.

giuseppe replied:
done

TomSweeneyRedHat commented on Atomic/rpmwriter.py:
```diff
@@ -0,0 +1,369 @@
+import os
```
Do you need to have a shebang here declaring the version of python you're supporting?  Perhaps:

#!/usr/bin/env python2

Given your python3 issue noted below?

giuseppe replied:
atomic must support both versions.  On RHEL we have python2, on Fedora python3

TomSweeneyRedHat replied:
Sounds good, I misunderstood your comment on the python3 failure.  Given that it would stlll be good to have some kind of shebang statement at the top of the file.  Perhaps:

#!/usr/bin/env python

But certainly not a comment to hold the merge for.

TomasTomecek replied:
What's the point of having shebang if the file is not meant to be run directly?

ashcrow commented on Atomic/rpm_host_install.py:
```diff
@@ -68,96 +70,60 @@ def rm_add_files_to_host(old_installed_files, exports, prefix="/", files_templat
+        rpm_name = "atomic-container-%s" % name
+        rpm_out = os.path.join(result_dir, "%s.rpm" % rpm_name)
+        def split_name_version(pkg):
+            r = r"([abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789\-\._\+]+)(.*)"
```
nit: This regex could be shortened to ```([a-zA-Z0-9_\-\.\+]+)(.*)```

giuseppe replied:
thanks, pushed a fixup patch.

ashcrow commented on Atomic/rpm_host_install.py:
```diff
@@ -199,10 +165,10 @@ def generate_rpm(name, image_id, labels, exports, destination, values=None, inst
             rpm_file = RPMHostInstall.find_rpm(rpm_root)
             if rpm_file:
-                dest_path = os.path.join(destination, "container.rpm")
+                orig_name = "atomic-container-%s.rpm" % name
```
nit: I recommend using ``{}`` and ``format`` with string formatting unless there is a need to hold compatibility with very old Python versions.

giuseppe replied:
change included in the fixup patch

ashcrow commented on Atomic/rpmwriter.py:
```diff
@@ -0,0 +1,369 @@
+    def add_conflict(self, name, version):
+        self.conflict.append([name, version])
+
+    def __init__(self, out, root, name, version, release, summary='', description='', license_='gpl2', changelog='', url='', group='', stderr=sys.stderr, whitelist=None):
```
I ❤️ GPL but why default to gpl2 if no license is provided?

giuseppe replied:
rpm complains if no license is specified, so I've just used the same license of `atomic` itself

TomSweeneyRedHat commented:
LGTM, assuming happy test results.  Thanks for the nit touch ups.

giuseppe commented:
@rhatdan what do you think of this change?

rhatdan commented:
@giuseppe Can this handle dependencies?  IE Can I build a "spec" file that says it requires "foobar" package be installed?

giuseppe commented:
@rhatdan the rpmwriter doesn't use a spec file to build the `.rpm` file. It creates the cpio archive given a `rootfs` and the metadata is read from labels and injected in the `.rpm`:

`LABEL atomic.requires=foo>1,bar<2`

We support: `summary`, `version`, `release`, `license`, `url`, `requires`, `provides`, `conflicts`, `description`.

I'll document this once the PR looks fine.

@TomasTomecek

TomasTomecek commented on Atomic/rpmwriter.py:
```diff
@@ -0,0 +1,369 @@
+# The complete list is available here:
+# https://github.com/rpm-software-management/rpm/blob/master/lib/rpmtag.h
+
+class RpmWriter(object):
```
My opinion is that this module should be a separate project. If you look at it, it's not tight to `atomic` in any way. I think that plenty of other users and projects could benefit from a lightweight and friendler user-experience interaction with rpm builds. Even further, it would be awesome if rpm team reviewed this piece of code and ack'd it.

cgwalters replied:
We're not *building* here though.  More like just repacking binaries.  What other use cases are there where one isn't using a spec file?  Or without wanting a dependency on `rpmbuild`?

giuseppe replied:
I've created a separate repository here: https://github.com/giuseppe/tinyrpmbuild, if someone is really interested in it.

I agree with @cgwalters, this is not a replacement for `rpmbuild`, it is really implementing the minimum necessary to get an installable `.rpm` file.  Not sure how many other users will have the same requirement of building an `.rpm` on the fly without using `rpmbuild`.

TomasTomecek commented on Atomic/rpmwriter.py:
```diff
@@ -0,0 +1,369 @@
+        def try_read(stdout, gzip_out):
+            written = 0
+            while True:
+                readable, _, _ = select.select([stdout], [], [], 0)
```
Wow, is `select` needed here? I would assume that simple pipe with reads and writes is just enough.

giuseppe replied:
I want it to not block on `_try_read`, just read as much as it is available.  In this way I can feed more data to the cpio process through `cpio_process.stdin.write(filename.encode() + b"\n")`

TomasTomecek commented on Atomic/rpmwriter.py:
```diff
@@ -0,0 +1,369 @@
+import shutil
+
+# More information on the RPM file format can be found here:
+# http://ftp.rpm.org/max-rpm/s1-rpm-file-format-rpm-file-format.html
```
Is this spec backwards and forward compatible?

baude commented:
@rh-atomic-bot r+  	8537474

baude commented:
@rh-atomic-bot r+ 271ebdb

giuseppe commented:
@baude thanks for the r+, I've manually fixed the conflicts with the fixup patches.  Pushed a new version ⬆️

rhatdan commented:
@rh-atomic-bot retry

rhatdan commented:
@rh-atomic-bot r+  	be51f9c
